subsys/bluetooth/controller/ll_sw/ull_llcp_conn.c

```diff
@@ -0,0 +1,766 @@
+/*
+ * Copyright (c) 2020 Demant
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/types.h>
+
+#include <bluetooth/hci.h>
+#include <sys/byteorder.h>
+#include <sys/slist.h>
+#include <sys/util.h>
+
+#include "hal/ccm.h"
+
+#include "util/mem.h"
+#include "util/memq.h"
+
+#include "pdu.h"
+#include "ll.h"
+#include "ll_settings.h"
+
+#include "lll.h"
+#include "lll_conn.h"
+
+#include "ull_conn_types.h"
+#include "ull_tx_queue.h"
+#include "ull_llcp.h"
+#include "ull_llcp_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_ctlr_ull_llcp
+#include "common/log.h"
+#include <soc.h>
+#include "hal/debug.h"
+
+/* LLCP Local Procedure Connection Update FSM states */
+enum {
+	LP_CU_STATE_IDLE,
+	LP_CU_STATE_WAIT_TX_CONN_PARAM_REQ,
+	LP_CU_STATE_WAIT_RX_CONN_PARAM_RSP,
+	LP_CU_STATE_WAIT_TX_CONN_UPDATE_IND,
+	LP_CU_STATE_WAIT_RX_CONN_UPDATE_IND,
+	LP_CU_STATE_WAIT_INSTANT,
+	LP_CU_STATE_WAIT_NTF,
+};
+
+/* LLCP Local Procedure Connection Update FSM events */
+enum {
+	/* Procedure run */
+	LP_CU_EVT_RUN,
+
+	/* Response recieved */
+	LP_CU_EVT_CONN_PARAM_RSP,
+
+	/* Indication recieved */
+	LP_CU_EVT_CONN_UPDATE_IND,
+
+	/* Reject response recieved */
+	LP_CU_EVT_REJECT,
+
+	/* Unknown response recieved */
+	LP_CU_EVT_UNKNOWN,
+};
+
+/* LLCP Remote Procedure Connection Update FSM states */
+enum {
+	RP_CU_STATE_IDLE,
+	RP_CU_STATE_WAIT_RX_CONN_PARAM_REQ,
+	RP_CU_STATE_WAIT_NTF_CONN_PARAM_REQ,
+	RP_CU_STATE_WAIT_CONN_PARAM_REQ_REPLY,
+	RP_CU_STATE_WAIT_TX_REJECT_IND,
+	RP_CU_STATE_WAIT_TX_CONN_PARAM_RSP,
+	RP_CU_STATE_WAIT_TX_CONN_UPDATE_IND,
+	RP_CU_STATE_WAIT_RX_CONN_UPDATE_IND,
+	RP_CU_STATE_WAIT_INSTANT,
+	RP_CU_STATE_WAIT_NTF,
+};
+
+/* LLCP Remote Procedure Connection Update FSM events */
+enum {
+	/* Procedure run */
+	RP_CU_EVT_RUN,
+
+	/* Request recieved */
+	RP_CU_EVT_CONN_PARAM_REQ,
+
+	/* Indication recieved */
+	RP_CU_EVT_CONN_UPDATE_IND,
+
+	/* CONN_PARAM_REQ reply */
+	RP_CU_EVT_CONN_PARAM_REQ_REPLY,
+
+	/* CONN_PARAM_REQ negative reply */
+	RP_CU_EVT_CONN_PARAM_REQ_NEG_REPLY,
+};
+
+/*
+ * ULL -> LL Interface
+ */
+
+extern void ll_rx_enqueue(struct node_rx_pdu *rx);
+
+/*
+ * LLCP Local Procedure Connection Update FSM
+ */
+
+static uint16_t lp_event_counter(struct ull_cp_conn *conn)
+{
+	/* TODO(thoh): Mocked lll_conn */
+	struct mocked_lll_conn *lll;
+	uint16_t event_counter;
+
+	/* TODO(thoh): Lazy hardcoded */
+	uint16_t lazy = 0;
+
+	/**/
+	lll = &conn->lll;
+
+	/* Calculate current event counter */
+	event_counter = lll->event_counter + lll->latency_prepare + lazy;
```

Comment:
Sidenote: I am missing the initialisation of event-counter. In the topic-ble-llcp branch initialisation is done, so this will fix itself automatically